File: android/app/src/main/java/fr/smarthomeworld/wealth/ui/Charts.kt
package fr.smarthomeworld.wealth.ui

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.gestures.detectHorizontalDragGestures
import androidx.compose.foundation.gestures.detectTapGestures
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.layout.onSizeChanged
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.IntOffset
import kotlin.math.roundToInt
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.size
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.unit.dp

/**
 * Two charts, drawn by hand.
 *
 * A charting library would be four megabytes and a theme of its own for
 * a line and a ring; Canvas draws both in fifty lines and inherits the
 * app's colours. No axes, no gridlines, no legend inside the picture:
 * the numbers that matter are written next to it as text, where a
 * screen reader can find them.
 */

/** The eight colours a share is drawn in, in order. */
val Slices = listOf(
    Color(0xFF4F86F7), Color(0xFF34D399), Color(0xFFF59E0B), Color(0xFFF87171),
    Color(0xFFA78BFA), Color(0xFF22D3EE), Color(0xFFFB7185), Color(0xFF94A3B8),
)

/**
 * The line of a value over time. Rising or falling decides the colour,
 * because the first thing anybody asks of this chart is the direction.
 *
 * A finger on the line picks the nearest day: a thin rule marks it and
 * a label above says when and how much. Sliding sideways walks through
 * time; lifting the finger puts the chart back as it was. Sideways
 * only, so a vertical swipe still scrolls the page the chart sits in.
 */
@Composable
fun LineChart(
    values: List<Double>,
    modifier: Modifier = Modifier.fillMaxWidth().height(170.dp),
    /** One ISO date per value, for the label; without them the chart
     *  still draws but a touch shows nothing. */
    dates: List<String> = emptyList(),
    currency: String = "EUR",
) {
    val up = values.size < 2 || values.last() >= values.first()
    val line = if (up) Gain else Loss
    val grid = MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.18f)
    val scrubbable = values.size > 1 && dates.size == values.size
    var picked by remember(values) { mutableStateOf<Int?>(null) }
    var width by remember { mutableStateOf(0) }

    fun indexAt(x: Float): Int {
        if (width <= 0) return 0
        val i = (x / width * (values.size - 1)).roundToInt()
        return i.coerceIn(0, values.lastIndex)
    }

    val touch = if (!scrubbable) Modifier else Modifier
        .pointerInput(values) {
            detectTapGestures(onPress = { at ->
                picked = indexAt(at.x)
                tryAwaitRelease()
                picked = null
            })
        }
        .pointerInput(values) {
            detectHorizontalDragGestures(
                onDragStart = { at -> picked = indexAt(at.x) },
                onDragEnd = { picked = null },
                onDragCancel = { picked = null },
                onHorizontalDrag = { change, _ ->
                    change.consume()
                    picked = indexAt(change.position.x)
                },
            )
        }

    Box(modifier.onSizeChanged { width = it.width }.then(touch)) {
        Canvas(Modifier.fillMaxSize()) {
            if (values.size < 2) return@Canvas
            val lo = values.min()
            val hi = values.max()
            val span = (hi - lo).takeIf { it > 0.0 } ?: 1.0
            val stepX = size.width / (values.size - 1)
            // A little air above and below, so the line never touches the edge.
            val top = size.height * 0.08f
            val usable = size.height * 0.84f

            fun pointAt(i: Int): Offset {
                val y = top + usable * (1f - ((values[i] - lo) / span).toFloat())
                return Offset(stepX * i, y)
            }

            drawLine(grid, Offset(0f, size.height), Offset(size.width, size.height), 1f)

            val path = Path().apply {
                moveTo(0f, pointAt(0).y)
                for (i in 1 until values.size) {
                    val p = pointAt(i)
                    lineTo(p.x, p.y)
                }
            }
            // The same path closed downwards, filled with a fading wash.
            val under = Path().apply {
                addPath(path)
                lineTo(size.width, size.height)
                lineTo(0f, size.height)
                close()
            }
            drawPath(under, Brush.verticalGradient(
                listOf(line.copy(alpha = 0.28f), Color.Transparent)))
            drawPath(path, line, style = Stroke(width = 2.5.dp.toPx(), cap = StrokeCap.Round))

            val i = picked
            if (i != null) {
                val p = pointAt(i)
                drawLine(line.copy(alpha = 0.5f), Offset(p.x, 0f), Offset(p.x, size.height),
                    1.dp.toPx())
                drawCircle(line, radius = 5.dp.toPx(), center = p)
            } else {
                drawCircle(line, radius = 4.dp.toPx(), center = pointAt(values.lastIndex))
            }
        }

        val i = picked
        if (i != null) {
            // The label rides above the finger but stays inside the
            // chart, so the first and last days are readable too.
            var labelWidth by remember { mutableStateOf(0) }
            val x = if (values.size > 1) width * i / (values.size - 1) else 0
            val left = (x - labelWidth / 2).coerceIn(0, (width - labelWidth).coerceAtLeast(0))
            Surface(
                Modifier
                    .offset { IntOffset(left, 0) }
                    .onSizeChanged { labelWidth = it.width },
                shape = RoundedCornerShape(8.dp),
                color = MaterialTheme.colorScheme.surfaceVariant,
                shadowElevation = 2.dp,
            ) {
                Column(Modifier.padding(horizontal = 10.dp, vertical = 6.dp)) {
                    Text(Fmt.day(dates[i]), style = MaterialTheme.typography.labelSmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant)
                    Text(Fmt.money(values[i], currency), style = MaterialTheme.typography.labelLarge,
                        fontWeight = FontWeight.SemiBold)
                }
            }
        }
    }
}

/**
 * The ring: one arc per share, biggest first, in the order the colours
 * are listed so that the legend beside it can use the same index.
 */
@Composable
fun DonutChart(
    shares: List<Double>,
    modifier: Modifier = Modifier.size(150.dp),
    centre: @Composable () -> Unit = {},
) {
    val track = MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.12f)
    Box(modifier, contentAlignment = Alignment.Center) {
        Canvas(Modifier.fillMaxSize()) {
            val total = shares.sum().takeIf { it > 0 } ?: 1.0
            val width = 18.dp.toPx()
            val inset = width / 2
            val box = Size(size.width - width, size.height - width)
            drawArc(track, 0f, 360f, useCenter = false,
                topLeft = Offset(inset, inset), size = box,
                style = Stroke(width = width))
            var start = -90f
            shares.forEachIndexed { i, share ->
                val sweep = (share / total * 360.0).toFloat()
                drawArc(
                    Slices[i % Slices.size], start, sweep - 1.5f, useCenter = false,
                    topLeft = Offset(inset, inset), size = box,
                    style = Stroke(width = width, cap = StrokeCap.Butt))
                start += sweep
            }
        }
        centre()
    }
}

File: ios/Wealth/UI/Charts.swift
import SwiftUI

// Two charts, drawn by hand, as on Android.
//
// A charting framework brings its own axes, legends and theme for a line
// and a ring; a `Path` draws both in a few lines and inherits the app's
// colours. No axes, no gridlines, no legend inside the picture: the
// numbers that matter are written next to it as text, where VoiceOver
// can find them.

/// The eight colours a share is drawn in, in order — the same as Android's.
let slices: [Color] = [0x4F86F7, 0x34D399, 0xF59E0B, 0xF87171,
                       0xA78BFA, 0x22D3EE, 0xFB7185, 0x94A3B8].map { rgb in
    Color(red: Double((rgb >> 16) & 0xFF) / 255,
          green: Double((rgb >> 8) & 0xFF) / 255,
          blue: Double(rgb & 0xFF) / 255)
}

/// The line of a value over time. Rising or falling decides the colour,
/// because the first thing anybody asks of this chart is the direction.
///
/// A short press on the line picks the nearest day: a thin rule marks it
/// and a label above says when and how much. Sliding then walks through
/// time; lifting the finger puts the chart back as it was. The press
/// comes first so that a swipe over the chart still scrolls the list.
struct LineChart: View {
    let values: [Double]
    /// One ISO date per value, for the label; without them the chart
    /// still draws but a press shows nothing.
    var dates: [String] = []
    var currency: String? = nil

    @State private var picked: Int?
    @State private var labelWidth: CGFloat = 0

    private var scrubbable: Bool { values.count > 1 && dates.count == values.count }

    var body: some View {
        let up = values.count < 2 || (values.last ?? 0) >= (values.first ?? 0)
        let colour: Color = up ? .gain : .loss
        GeometryReader { geo in
            let size = geo.size
            if values.count > 1 {
                let points = points(in: size)
                ZStack(alignment: .topLeading) {
                    // The same line closed downwards, filled with a fading wash.
                    Path { p in
                        p.move(to: CGPoint(x: 0, y: size.height))
                        points.forEach { p.addLine(to: $0) }
                        p.addLine(to: CGPoint(x: size.width, y: size.height))
                        p.closeSubpath()
                    }
                    .fill(LinearGradient(colors: [colour.opacity(0.28), .clear],
                                         startPoint: .top, endPoint: .bottom))
                    Path { p in p.addLines(points) }
                        .stroke(colour, style: StrokeStyle(lineWidth: 2.5, lineCap: .round, lineJoin: .round))
                    if let i = picked, points.indices.contains(i) {
                        Rectangle().fill(colour.opacity(0.5))
                            .frame(width: 1, height: size.height)
                            .offset(x: points[i].x - 0.5)
                        Circle().fill(colour)
                            .frame(width: 10, height: 10)
                            .position(points[i])
                    } else if let last = points.last {
                        Circle().fill(colour)
                            .frame(width: 8, height: 8)
                            .position(last)
                    }
                    Rectangle().fill(Color.secondary.opacity(0.18))
                        .frame(width: size.width, height: 1)
                        .offset(y: size.height - 1)
                    if let i = picked, points.indices.contains(i) {
                        label(i)
                            .offset(x: min(max(points[i].x - labelWidth / 2, 0),
                                           max(size.width - labelWidth, 0)))
                    }
                }
                .contentShape(Rectangle())
                .gesture(scrub(width: size.width), including: scrubbable ? .all : .subviews)
            }
        }
        .accessibilityHidden(true)
    }

    /// The label rides above the finger but stays inside the chart, so
    /// the first and last days are readable too.
    private func label(_ i: Int) -> some View {
        VStack(alignment: .leading, spacing: 1) {
            Text(Fmt.day(dates[i])).font(.caption2).foregroundStyle(.secondary)
            Text(Fmt.money(values[i], currency)).font(.subheadline.weight(.semibold))
        }
        .padding(.horizontal, 10).padding(.vertical, 6)
        .background(.regularMaterial, in: RoundedRectangle(cornerRadius: 8))
        .fixedSize()
        .background(GeometryReader { g in
            Color.clear.onAppear { labelWidth = g.size.width }
                .onChange(of: g.size.width) { _, w in labelWidth = w }
        })
    }

    private func scrub(width: CGFloat) -> some Gesture {
        LongPressGesture(minimumDuration: 0.15)
            .sequenced(before: DragGesture(minimumDistance: 0))
            .onChanged { value in
                switch value {
                case .first(true):
                    // The press is recognised before the finger moves;
                    // the drag's first event says where it is.
                    break
                case .second(true, let drag?):
                    let i = index(at: drag.location.x, width: width)
                    if i != picked {
                        picked = i
                        UISelectionFeedbackGenerator().selectionChanged()
                    }
                default:
                    break
                }
            }
            .onEnded { _ in picked = nil }
    }

    private func index(at x: CGFloat, width: CGFloat) -> Int {
        guard width > 0, values.count > 1 else { return 0 }
        let i = Int((x / width * CGFloat(values.count - 1)).rounded())
        return min(max(i, 0), values.count - 1)
    }

    private func points(in size: CGSize) -> [CGPoint] {
        let lo = values.min() ?? 0, hi = values.max() ?? 0
        let span = hi > lo ? hi - lo : 1
        let step = size.width / CGFloat(values.count - 1)
        // A little air above and below, so the line never touches the edge.
        let top = size.height * 0.08, usable = size.height * 0.84
        return values.enumerated().map { i, v in
            CGPoint(x: step * CGFloat(i), y: top + usable * (1 - CGFloat((v - lo) / span)))
        }
    }
}

/// The ring: one arc per share, biggest first, in the order the colours
/// are listed so that the legend beside it can use the same index.
struct DonutChart<Centre: View>: View {
    let shares: [Double]
    var width: CGFloat = 18
    @ViewBuilder var centre: () -> Centre

    var body: some View {
        let total = max(shares.reduce(0, +), .leastNonzeroMagnitude)
        ZStack {
            Circle().stroke(Color.secondary.opacity(0.12), lineWidth: width)
            ForEach(Array(shares.enumerated()), id: \.offset) { i, share in
                let start = shares.prefix(i).reduce(0, +) / total
                let end = start + share / total
                // A hair's gap between slices, unless there is only one.
                let gap = shares.count > 1 ? min(0.004, (end - start) / 2) : 0
                Circle()
                    .trim(from: start, to: max(start, end - gap))
                    .stroke(slices[i % slices.count], style: StrokeStyle(lineWidth: width, lineCap: .butt))
                    .rotationEffect(.degrees(-90))
            }
            centre()
        }
        .padding(width / 2)
    }
}
